k2/python/k2/fsa.py

```diff
@@ -644,33 +656,35 @@ def get_arc_post(self, use_double_scores: bool,
         '''Compute scores on arcs, representing log probabilities;
         with log_semiring=True you could call these log posteriors,
         but if log_semiring=False they can only be interpreted as the
-        difference betwen the best-path score and the score of the
+        difference between the best-path score and the score of the
         best path that includes this arc.
-        This version is differentiable; see also _get_arc_post().
-        Caution: because of how the autograd mechanics works and the
-        need to avoid circular references, this is not cached; it's
-        best to store it if you'll need it multiple times.
+        This version is differentiable; see also :func:`_get_arc_post`.
+
+        Caution:
+          Because of how the autograd mechanics works and the
+          need to avoid circular references, this is not cached;
+          it's best to store it if you'll need it multiple times.
 
         Args:
           use_double_scores: if True, use double precision.
           log_semiring: if True, use log semiring, else tropical.
-        Returns: a torch.Tensor with shape equal to (num_arcs,)
+        Returns:
+          a torch.Tensor with shape equal to (num_arcs,)
           and non-positive elements.
         '''
         # We don't cache this!  User should store it if needed more than once,
         # to avoid duplicate code in backprop.  We may be able to partially fix
         # this at some point with a weak dictionary.
-        forward_scores = self._get_forward_scores(use_double_scores,
-                                                  log_semiring)
-        backward_scores = self._get_backward_scores(use_double_scores,
-                                                    log_semiring)
+        forward_scores = self.get_forward_scores(use_double_scores,
```

Comment:
An issue for later is to try to use a weak dictionary to introduce a caching mechanism for the results of  get_forward_scores and get_backward_scores and get_arc_post (and similar differentiable things), so we don't end up doing the backprop multiple times in cases where, say, the forward and backward scores are used multiple times (which will be common).

Reply:
Created an issue for it https://github.com/k2-fsa/k2/issues/601